File: xapian-core/bindings/java/ApiTest.java
/* ApiTest.java: class containing java implementations of apitest tests.
 *
 * ----START-LICENCE----
 * Copyright 1999,2000 BrightStation PLC
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License as
 * published by the Free Software Foundation; either version 2 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307
 * USA
 * -----END-LICENCE-----
 */
import com.muscat.om.*;
import java.util.*;
import java.io.*;

public class ApiTest {
    public static void main(String[] args) throws Throwable {
        backendmanager = new BackendManager();

	backendmanager.set_datadir(System.getProperty("srcdir", ".") +
	                           "/../../tests/testdata/" );

        System.out.println("pctcutoff1: " + test_pctcutoff1());
        System.out.println("collapsekey1: " + test_collapsekey1());
    }

    public static boolean test_pctcutoff1() throws Throwable
    {
	boolean success = true;

	OmEnquire enquire = new OmEnquire(get_simple_database());
	init_simple_enquire(enquire);

	OmMSet mymset1 = enquire.get_mset(0, 100);

	if (verbose) {
	    System.out.println("MSet: " + mymset1.get_description());
	    System.out.println("Original mset pcts:");
	    print_mset_percentages(mymset1);
	    System.out.println("");
	}

	int num_items = 0;
	int my_pct = 100;
	int changes = 0;
	OmVector mymset1_items = mymset1.get_items();
	for (int i=0; i<mymset1_items.size(); ++i) {
	    int new_pct = mymset1.convert_to_percent((OmMSetItem)mymset1_items.elementAt(i));
	    if (new_pct != my_pct) {
		changes++;
		if (changes <= 3) {
		    num_items = i;
		    my_pct = new_pct;
		}
	    }
	}

	if (changes <= 3) {
	    success = false;
	    if (verbose) {
		System.out.println("MSet not varied enough to test");
	    }
	}
	if (verbose) {
	    System.out.println("Cutoff percent: " + my_pct);
	}

	OmMatchOptions mymopt = new OmMatchOptions();
	mymopt.set_percentage_cutoff(my_pct);
	OmMSet mymset2 = enquire.get_mset(0, 100, null, mymopt);
	OmVector mymset2_items = mymset2.get_items();

	if (verbose) {
	    System.out.println("Percentages after cutoff:");
	    print_mset_percentages(mymset2);
	    System.out.println("");
	}

	if (mymset2_items.size() < num_items) {
	    success = false;
	    if (verbose) {
		System.out.println("Match with % cutoff lost too many items");
	    }
	} else if (mymset2_items.size() > num_items) {
	    for (int i=num_items; i<mymset2_items.size(); ++i) {
		if (mymset2.convert_to_percent((OmMSetItem)mymset2_items.elementAt(i)) != my_pct) {
		    success = false;
		    if (verbose) {
			System.out.println("Match with % cutoff returned "
				+ " too many items");
		    }
		    break;
		}
	    }
	}

	return success;
    }

    public static boolean test_collapsekey1() throws Throwable
    {
	boolean success = true;

	OmEnquire enquire = new OmEnquire(get_simple_database());
	init_simple_enquire(enquire);

	OmMatchOptions mymopt = new OmMatchOptions();

	OmMSet mymset1 = enquire.get_mset(0, 100, 0, mymopt);
	mymset1_items = mymset1.get_items();
	int mymsize1 = mymset1_items.size();

	for (int key_no = 1; key_no<7; ++key_no) {
	    mymopt.set_collapse_key(key_no);
	    OmMSet mymset = enquire.get_mset(0, 100, 0, mymopt);
	    mymset_items = mymset.get_items();

	    if(mymsize1 <= mymset_items.size()) {
		success = false;
		if (verbose) {
		    System.out.println("Had no fewer items when performing collapse: don't know whether it worked.");
		}
	    }

	    map<string, om_docid> keys;
	    for (vector<OmMSetItem>::const_iterator i=mymset.items.begin();
		 i != mymset.items.end();
		 ++i) {
		OmKey key = enquire.get_doc(*i).get_key(key_no);
		if (key.value != i->collapse_key.value) {
		    success = false;
		    if (verbose) {
			cout << "Expected key value was not found in MSetItem: "
				<< "expected `" << key.value
				<< "' found `" << i->collapse_key.value
				<< "'" << endl;

		    }
		}
		if (keys[key.value] != 0 && key.value != "") {
		    success = false;
		    if (verbose) {
			cout << "docids " << keys[key.value]
				<< " and " << i->did
				<< " both found in MSet with key `" << key.value
				<< "'" << endl;
		    }
		    break;
		} else {
		    keys[key.value] = i->did;
		}
	    }
	    // don't bother continuing if we've already failed.
	    if (!success) break;
	}

	return success;
    }

    public static OmDatabaseGroup get_simple_database() throws Throwable {
        OmDatabase mydb = get_database("apitest_simpledata");
	return make_dbgrp(mydb);
    }

    public static void init_simple_enquire(OmEnquire enq) throws Throwable {
        enq.set_query(new OmQuery("thi"));
    }

    public static void print_mset_percentages(OmMSet mset)
    {
        OmVector mset_items = mset.get_items();
        for (int i=0; i<mset_items.size(); ++i) {
	    System.out.print(" ");
	    System.out.print(mset.convert_to_percent(
	                     (OmMSetItem)mset_items.elementAt(i)));
	}
    }

    public static boolean verbose = true;

    private static BackendManager backendmanager;

    private static OmDatabase get_database(String dbname) throws Throwable {
        return backendmanager.get_database(dbname);
    }

    private static OmDatabaseGroup make_dbgrp(OmDatabase db1) throws Throwable {
        OmDatabaseGroup result = new OmDatabaseGroup();

	result.add_database(db1);

	return result;
    }
}

File: xapian-core/bindings/python/apitest.py
#!/usr/bin/python
from omuscat import *

def get_simple_database():
    mydb = OmDatabaseGroup()
    mydb.add_database("sleepycat", ["/home/cemerson/working/open-muscat/build/om-debug-valis/tests/.sleepy/db=apitest_simpledata="])
    return mydb

def init_simple_enquire(enq, query = OmQuery("thi")):
    enq.set_query(query)

def do_get_simple_query_mset(query, maxitems = 10, first = 0):
    enquire = OmEnquire(get_simple_database())
    init_simple_enquire(enquire, query)

    return enquire.get_mset(first, maxitems)

def print_mset_percentages(mset):
    items = mset.items
    for i in range(len(items)):
        print " ",
	print mset.convert_to_percent(items[i][1]),

def test_pctcutoff1():
    success = 1

    enquire = OmEnquire(get_simple_database())
    init_simple_enquire(enquire)

    mymset1 = enquire.get_mset(0, 100)

    print "Original mset pcts:",
    print_mset_percentages(mymset1)
    print

    num_items = 0;
    my_pct = 100;
    changes = 0;
    mymset1items = mymset1.items
    for i in range(len(mymset1items)):
        new_pct = mymset1.convert_to_percent(mymset1items[i][1]);
        if (new_pct != my_pct):
	    changes = changes + 1
	    if (changes <= 3):
	        num_items = i
		my_pct = new_pct

    if (changes <= 3):
	success = 0
	print "MSet not varied enough to test"

    print "Cutoff percent: ", my_pct

    mymopt = OmMatchOptions()
    mymopt.set_percentage_cutoff(my_pct)
    mymset2 = enquire.get_mset(0, 100, "NULL", mymopt)

    print "Percentages after cutoff:",
    print_mset_percentages(mymset2)
    print

    mymset2items = mymset2.items
    if (len(mymset2items) < num_items):
        success = 0
	print "Match with % cutoff lost too many items"
    elif (len(mymset2items) > num_items):
        for i in range(num_items, len(mymset2items)):
	    if (mymset2.convert_to_percent(mymset2.items[i]) != my_pct):
	        success = 0
		print "Match with % cutoff returned  too many items"
		break

    return success

def test_collapsekey1():
    success = 1

    enquire = OmEnquire(get_simple_database())
    init_simple_enquire(enquire)

    mymopt = OmMatchOptions()
    mymset1 = enquire.get_mset(0, 100, "NULL", mymopt);
    mymsize1 = len(mymset1.items)

    for key_no in range(1, 8):
	mymopt.set_collapse_key(key_no)
	mymset = enquire.get_mset(0, 100, "NULL", mymopt)

	if(mymsize1 <= len(mymset.items)):
	    success = 0;
	    print "Had no fewer items when performing collapse: don't know whether it worked."

	keys = {}
	for i in mymset.items:
	    key = enquire.get_doc(i[0]).get_key(key_no)
	    if (key != i[2]):
		success = 0;
		print "Expected key value was not found in MSetItem: ",
		print "expected `", key,
		print "' found `", i[2],
		print "'"

	    if (keys.has_key(key) and keys[key] != 0 and key != ""):
	        success = 0;
		print "docids", keys[key], "and", i[0], "both found in MSet with key `", key, "'"
		break
	    else:
	        keys[key] = i[0]
	if (not success):
	    break

    return success

def test_expandmaxitems1():
    enquire = OmEnquire(get_simple_database())
    init_simple_enquire(enquire)

    mymset = enquire.get_mset(0, 10)

    if (len(mymset.items) < 2):
        return 0;

    myrset = OmRSet()
    mymsetitems = mymset.items
    myrset.add_document(mymsetitems[0][0])
    myrset.add_document(mymsetitems[1][0])

    myeset = enquire.get_eset(1, myrset)

    return (len(myeset.items) == 1)

def test_batchquery1():
    success = 1
    verbose = 1

    myqueries = [
	( OmQuery("thi"), 0, 10, 0, 0, 0),
	( OmQueryNull(), 0, 10, 0, 0, 0),
	( OmQuery("word"), 0, 10, 0, 0, 0)]

    print myqueries

    benq = OmBatchEnquire(get_simple_database())

    print "foo"
    benq.set_queries(myqueries)

    print "bar"
    myresults = benq.get_msets()

    print "myresults = ", myresults
    if (len(myresults) != 3):
	success = 0;
	if (verbose):
	    print "Results size is", len(myresults), "expected 3."
	    print "result is: ", myresults
    else:
	if (myresults[0] != do_get_simple_query_mset(OmQuery("thi"))):
	    success = 0
	    if (verbose):
		print "Query 1 returned different result!"
		print "myresults[0] =", myresults[0]
		print "simple =", do_get_simple_query_mset(OmQuery("thi"))
	try:
 	    foo = myresults[1]
	    success = 0
	    if (verbose):
		print "Query 2 should not be valid"
	except:
	    pass

	if (myresults[2] != do_get_simple_query_mset(OmQuery("word"))):
	    success = 0
	    if (verbose):
		print "Query 3 returned different result!"

    return success
